### PR TITLE
feature: support kryo codec

### PR DESCRIPTION
**Pull request by jsbxyyx**
protocol kryo serializer

### CODE REVIEW

**Comment by ujjboy** [comment]
Duplicate with #1418, can you close one PR?

**Comment by ujjboy** [comment]
@jsbxyyx  sorry, I didn't notice that.

File: codec/seata-codec-kryo/src/main/java/io/seata/codec/kryo/KryoCodec.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+    @Override
+    public <T> T decode(byte[] bytes) {
+        if (bytes == null || bytes.length == 0) {
+            throw new IllegalArgumentException("bytes is null");
+        }
+        KryoSerializer kryoSerializer = KryoSerializerFactory.getInstance().get();
+        try {
+            T t = kryoSerializer.deserialize(bytes);
+            return t;
```
Comment (xingfudeshi) [inline codec/seata-codec-kryo/src/main/java/io/seata/codec/kryo/KryoCodec.java]: return kryoSerializer.serialize(t);

File: codec/seata-codec-kryo/src/main/java/io/seata/codec/kryo/KryoCodec.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    @Override
+    public <T> byte[] encode(T t) {
+        if (t == null || !(t instanceof AbstractMessage)) {
+            throw new IllegalArgumentException("message is null");
+        }
+        KryoSerializer kryoSerializer = KryoSerializerFactory.getInstance().get();
+        try {
+            byte[] bytes = kryoSerializer.serialize(t);
+            return bytes;
```
Comment (xingfudeshi) [inline codec/seata-codec-kryo/src/main/java/io/seata/codec/kryo/KryoCodec.java]: return kryoSerializer.serialize(t);

File: codec/seata-codec-kryo/src/main/java/io/seata/codec/kryo/KryoSerializer.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    public Kryo getKryo() {
+        return kryo;
+    }
+
+    public <T> byte[] serialize(T t) {
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        Output output = new Output(baos);
+        kryo.writeClassAndObject(output, t);
+        output.close();
```
Comment (xingfudeshi) [inline codec/seata-codec-kryo/src/main/java/io/seata/codec/kryo/KryoSerializer.java]: Do we need flush?

File: codec/seata-codec-kryo/src/main/java/io/seata/codec/kryo/KryoSerializer.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        return baos.toByteArray();
+    }
+
+    public <T> T deserialize(byte[] bytes) {
+        ByteArrayInputStream bais = new ByteArrayInputStream(bytes);
+        Input input = new Input(bais);
+        input.close();
```
Comment (xingfudeshi) [inline codec/seata-codec-kryo/src/main/java/io/seata/codec/kryo/KryoSerializer.java]: Close is called before kryo.readClassAndObject(input);Is there a problem?
image

File: codec/seata-codec-kryo/src/main/java/io/seata/codec/kryo/KryoSerializer.java (hunk 3 above)
Comment (jsbxyyx) [reply codec/seata-codec-kryo/src/main/java/io/seata/codec/kryo/KryoSerializer.java]: close() inner call flush()

File: codec/seata-codec-kryo/src/main/java/io/seata/codec/kryo/KryoSerializer.java (hunk 4 above)
Comment (jsbxyyx) [reply codec/seata-codec-kryo/src/main/java/io/seata/codec/kryo/KryoSerializer.java]: FileInputStream close and ByteArrayInputStream close implements different, here not problem.

File: codec/seata-codec-kryo/src/main/java/io/seata/codec/kryo/KryoCodec.java (hunk 1 above)
Comment (jsbxyyx) [reply codec/seata-codec-kryo/src/main/java/io/seata/codec/kryo/KryoCodec.java]: ok

File: codec/seata-codec-kryo/src/main/java/io/seata/codec/kryo/KryoCodec.java (hunk 2 above)
Comment (jsbxyyx) [reply codec/seata-codec-kryo/src/main/java/io/seata/codec/kryo/KryoCodec.java]: ok

**Pull request merged**